src/workerd/api/eventsource.h

```diff
@@ -0,0 +1,176 @@
+#pragma once
+#include <workerd/jsg/jsg.h>
+#include <workerd/jsg/url.h>
+#include "basics.h"
+
+namespace workerd::api {
+
+using kj::uint;
+class Fetcher;
+
+// Implements the web standard EventSource API.
+class EventSource: public EventTarget {
+public:
+  class ErrorEvent final: public Event {
+  public:
+    ErrorEvent(jsg::Lock& js, const jsg::JsValue& error)
+        : Event(kj::str("error")),
+          error(js, error) {}
+
+    static jsg::Ref<ErrorEvent> constructor() = delete;
+    JSG_RESOURCE_TYPE(ErrorEvent) {
+      JSG_INHERIT(Event);
+      JSG_LAZY_READONLY_INSTANCE_PROPERTY(error, getError);
+    }
+
+  private:
+    jsg::JsRef<jsg::JsValue> error;
+
+    jsg::JsValue getError(jsg::Lock& js) { return error.getHandle(js); }
+  };
+
+  class OpenEvent final: public Event {
+  public:
+    OpenEvent() : Event(kj::str("open")) {}
+    static jsg::Ref<ErrorEvent> constructor() = delete;
+    JSG_RESOURCE_TYPE(OpenEvent) {
+      JSG_INHERIT(Event);
+    }
+  };
+
+  class MessageEvent final: public Event {
+  public:
+    explicit MessageEvent(kj::Maybe<kj::String>& type,
+                 kj::String data,
+                 kj::Maybe<kj::String> lastEventId,
+                 jsg::Url& url)
+        : Event(kj::mv(type).orDefault([] { return kj::str("message");})),
+          data(kj::mv(data)),
+          lastEventId(kj::mv(lastEventId)),
+          origin(url.getOrigin()) {}
+
+    static jsg::Ref<ErrorEvent> constructor() = delete;
+    JSG_RESOURCE_TYPE(MessageEvent) {
+      JSG_INHERIT(Event);
+      JSG_LAZY_READONLY_INSTANCE_PROPERTY(data, getData);
+      JSG_LAZY_READONLY_INSTANCE_PROPERTY(origin, getOrigin);
+      JSG_LAZY_READONLY_INSTANCE_PROPERTY(lastEventId, getLastEventId);
+    }
+
+  private:
+    kj::String data;
+    kj::Maybe<kj::String> lastEventId;
+    kj::Array<const char> origin;
+
+    kj::StringPtr getData() { return data; }
+    kj::Maybe<kj::StringPtr> getLastEventId() {
+      return lastEventId.map([](kj::String& s) { return s.asPtr(); });
+    }
+    kj::ArrayPtr<const char> getOrigin() { return origin; }
+  };
+
+  struct EventSourceInit {
+    // We don't actually make use of the standard withCredentials option. If this is set to
+    // any truthy value, we'll throw.
+    jsg::Optional<bool> withCredentials;
+
+    // This is a non-standard workers-specific extension that allows the EventSource to
```

Comment:
I wonder if the better extension would be to allow an `EventSource` to be constructed directly from a `Response`. It feels very weird that this interface is reproducing a lot of the `fetch()` interface.

Reply:
you mean like `new EventSource(response, ...)` ? or something else?

I view `EventSource` largely as just a convenience wrapper around fetch. One additional extension I was considering was making it so that we could wrap an `EventSource` around any `ReadableStream` so that we could use it even without performing a fetch at all... but I also don't want to stray too far from the standard API.

```
// I wouldn't want to diverge the EventSource constructor too far from the standard...
const eventSource = EventSource.from(readableStream);
```

Reply:
Yeah if EventSource doesn't actually depend on anything in the Response except for the body stream, then it really ought to be possible to create one from an arbitrary ReadableStream.

The fact that this API wants to make a fetch on your behalf makes it rather non-composable. Supporting fetcher bindings is one thing. But what about if you receive a Response (or a ReadableStream) over RPC, and want to parse that?

Also, I guess this API doesn't let you specify any custom headers on the request? Or see the headers on the response? This all seems very limiting for no good reason.

(I understand this is how the standard was written, just seems like a poorly-designed standard TBH.)

Reply:
heh, it's certainly not my favorite..

The most sophisticated bit is the auto-reconnection. This implementation does not currently do so, but there's a header that's supposed to be set in the reconnection fetch request to specify the last known event id provided by the server in a previous connection. Beyond that, the only thing it uses in the fetch is checking the response status code and verifying that the content-type is correct.

I think I will go ahead and add the non-standard `EventSource.from(...)` static method to make it possible to create an instance from a `ReadableStream` (or any async iterable). Then likely keep the rest as is.

Reply:
Added `EventSource.from(...)`